src/connector/src/sink/deltalake.rs

```diff
@@ -0,0 +1,544 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use deltalake::kernel::{Action, Add, DataType as DeltaLakeDataType, PrimitiveType, StructType};
+use deltalake::operations::transaction::commit;
+use deltalake::protocol::{DeltaOperation, SaveMode};
+use deltalake::table::builder::s3_storage_options::{
+    AWS_ACCESS_KEY_ID, AWS_ALLOW_HTTP, AWS_ENDPOINT_URL, AWS_REGION, AWS_S3_ALLOW_UNSAFE_RENAME,
+    AWS_SECRET_ACCESS_KEY,
+};
+use deltalake::writer::{DeltaWriter, RecordBatchWriter};
+use deltalake::DeltaTable;
+use risingwave_common::array::{to_record_batch_with_schema, StreamChunk};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+use risingwave_common::error::anyhow_error;
+use risingwave_common::types::DataType;
+use risingwave_common::util::iter_util::ZipEqFast;
+use risingwave_pb::connector_service::sink_metadata::Metadata::Serialized;
+use risingwave_pb::connector_service::sink_metadata::SerializedMetadata;
+use risingwave_pb::connector_service::SinkMetadata;
+use serde_derive::{Deserialize, Serialize};
+use serde_with::serde_as;
+use with_options::WithOptions;
+
+use super::coordinate::CoordinatedSinkWriter;
+use super::writer::{LogSinkerOf, SinkWriter};
+use super::{
+    Result, Sink, SinkCommitCoordinator, SinkError, SinkParam, SinkWriterParam,
+    SINK_TYPE_APPEND_ONLY, SINK_USER_FORCE_APPEND_ONLY_OPTION,
+};
+use crate::sink::writer::SinkWriterExt;
+
+pub const DELTALAKE_SINK: &str = "deltalake_rust";
+
+#[derive(Deserialize, Serialize, Debug, Clone, WithOptions)]
+pub struct DeltaLakeCommon {
```

Comment:
Do users need to set either `local.path` or `s3.path` to specify whether they are using s3 or local path? If so we'd better only let them specify a `path` and we check the prefix to tell if its using local path or s3.

Reply:
Using two options may better distinguish the differences between the two paths, and S3 requires additional options. Adding this prefix also reduces the likelihood of misunderstandings.

Reply:
I think it's more natural for users to set `path = s3:///path` or `path = file:///path` instead of `local.path = /path` or `s3.path = /path`. In this future when we support sinking to location type other than local file system and s3, it is unlikely for us to always add a `xxx.path` for each location type.